keps/sig-storage/20200120-skip-permission-change.md

```diff
@@ -0,0 +1,147 @@
+   different node.
+ - If volume has a large number of files, performing recursive `chown` and `chmod`
+   could be slow and could cause timeout while starting the pod.
+
+### Goals
+
+ - Allow volume ownership and permission to be skipped during mount
+
+### Non-Goals
+
+ - In some cases if user brings in a large enough volume from outside, the first time ownership and permission change still could take lot of time.
+ - On SELinux enabled distributions we will still do recursive chcon whenever applicable and handling that is outside the scope.
+
+## Proposal
+
+We propose that an user can optionally opt-in to skip recursive ownership(and permission) change on the volume if volume already has right permissions.
+
+### Implementation Details/Notes/Constraints [optional]
+
+When creating a pod, we propose that `pod.Spec.SecurityContext` field expanded to include a new field called `VolumePermissionChangePolicy` which can have following possible values:
+
+ - `NoChange` --> Don't change permissions and ownership. This could be useful to users when security policy of a cluster prevents users from removing `fsGroup` from their pod's `SecurityContext`.
+ - `Always` --> Always change the permissions and ownership to match fsGroup. This is the current behavior and it will be the default one when this proposal is implemented. Algorithm that performs permission change however will be changed to perform permission change of top level directory last.
+ - `OnDemand` --> Only perform permission change if permissions of top level directory does not match with expected permissions.
+
+```go
+type PodVolumePermissionChangePolicy string
+
+const(
+    NeverChangeVolumePermission PodVolumePermissionChangePolicy = "NoChange"
+    OnDemandChangeVolumePermission PodVolumePermissionChangePolicy = "OnDemand"
+    AlwaysChangeVolumePermission PodVolumePermissionChangePolicy = "Always"
+)
+
+type PodSecurityContext struct {
+    // VolumePermissionChangePolicy ← new field
+    // + optional
+    VolumePermissionChangePolicy *PodVolumePermissionChangePolicy
+}
+```
+
+### Risks and Mitigations
+
+- One of the risks is if user volume's permission was previously changed using old algorithm(which changes permission of top level directory first) and user opts in for `OnDemand` `VolumePermissionChangePolicy` then we can't distinguish if the volume was previously only partially recursively chown'd.
+
+
+## Graduation Criteria
+
+* Alpha in 1.18 provided all tests are passing and gated by the feature Gate
+   ConfigurableVolumeFilePermissions and set to a default of `False`
+
+* Beta in 1.19 with design validated by at least two customer deployments
+  (non-production), with discussions in SIG-Storage regarding success of
+  deployments.  A metric will be added to report time taken to perform a
+  volume ownership change.
+* GA in 1.20, with Node E2E tests in place tagged with feature Storage
```

Comment:
we should have e2es at the beta phase too

Reply:
done